### PR TITLE
Feature/new learning modes

### PR DESCRIPTION
**Pull request by ivan-pavlov**
Before building a new version of this package, please install the newest version of `libvw` (8.6.1).

After this PR I will focus on a vignette, demos, and bug fixes. And only after that return to `average loss` vector retrieving and plotting. And also there is a couple of features not stated in my GSoC proposal that I want to implement.

### CODE REVIEW

File: R/init.R
Comment (coatless) [inline R/init.R]: Let's move this out of `onAttach`, promote it to global status, and enclose inside of an environment. e.g. 

```r
# An environment to hold global variables and settings for rwv.
rwv_global <- new.env(parent = emptyenv())

rwv_global$general_check <- list(random_seed=0, ...)
rwv_global$check_lists <- list(general_check=rwv_global$general_check)
rwv_global$flatten_check_lists <- .flatten(rwv_global$check_lists)
```

Comment (ivan-pavlov) [reply R/init.R]: Yes, I will fix this.

Just for my own better understanding, why don't we want to do it in `onAttach`?

Comment (eddelbuettel) [reply R/init.R]: Should not really matter.  I tend to do it in `.onAttach()` myself and only have the `new.env()` call outside of it.

**Review by eddelbuettel — COMMENTED** [review]
> Before building a new version of this package, please install the newest version of libvw (8.6.1).

If you step back and look at this, was that a good idea with a week to go?

It set me back a few days until I had time to pick up the Debian package for 8.6.1 and rebuild it on Ubuntu 18.04 where I work. That is done.

But your project no longer builds.

That may be an error in the Debian packaging (some headers missing; I copied them for now into the sources of your package) and even once I do that it fails over "multiple definition"

I must presume that you also break the Travis if you genuinely depend on a version that Travis does not have.

So we are almost at the end and we do not have a building project.  

What do suggest we do now?

**Comment by eddelbuettel** [comment]
By also adjusting your `src/Makevars` to point to the include directory I have, and making sure no files are listed twice, I can build -- provided I also add these five files (which _may_ be missing due to a Debian packaging error, I do not know):

```
edd@rob:~/git/rvwgsoc(feature/new-learning-modes)$ ls src/extra/
error_reporting.h  example_predict.h  hash.h  no_label.h  parser_helper.h
edd@rob:~/git/rvwgsoc(feature/new-learning-modes)$ 
```

But I still don't understand how you expect Travis to get VW 8.6.1.

**Comment by eddelbuettel** [comment]
I pushed a new branch with my changes.  One suppresses a compiler warnings, it adds the headers missing here (for reasons we do not yet understand but may exogeneous to your project) plus a (crude !!) change to `src/Makevars`

So don't merge my branch as is.  This is more of a heads-up.

**Comment by coatless** [comment]
@eddelbuettel the missing headers seem to be an upstream issue with how their [Makefile.am](https://github.com/JohnLangford/vowpal_wabbit/blob/64ac2f8a1f566bb35c811f4f4d2fc8243b5dc5f6/Makefile.am#L3) omits the necessary headers used in [`global_data.h`](https://github.com/JohnLangford/vowpal_wabbit/blob/64ac2f8a1f566bb35c811f4f4d2fc8243b5dc5f6/vowpalwabbit/global_data.h#L19-L32). c.f. JohnLangford/vowpal_wabbit#1567

**Comment by eddelbuettel** [comment]
I had in the meantime reached out to the Debian maintainer but not yet heard back. I concur that is looks like an upstream issue so thanks for filing it there.  We'll see what they say.

**Comment by ivan-pavlov** [comment]
I see the problem. Thank you a lot for taking care of this issue! 
In case there won't be a quick solution with version 8.6.1 I can easily roll back to 8.5.0. Essentially the only thing that's need to be changed is the list of acceptable arguments. So I can prepare a small and easy switch, that will choose between different acceptable arguments lists depending on an installed version of `VW`. 

There's one thing that I don't fully understand yet. How can we determine correct include directory with` Rcpp`?
`src/Makevars`:
```
-PKG_CPPFLAGS = -I/usr/local/include/
+## we need a configure snipped to determine the right include directory
+PKG_CPPFLAGS = -I/usr/include/vowpalwabbit -Iextra/
```
 Can we use the help of configure scripts here?

**Comment by eddelbuettel** [comment]
I agree that 8.6.1 looks more appealing.  

Configure is work. We only have this week --> don't do it. I suggest a shell script (or R script) called `configure` which simple loops over a known set of location `/usr/local/include`, `/usr/local/include/vowpalwabbit`, `/usr/local/include/vw`, ... (ditto for `/usr`, `/opt`, ...) until it finds a plausible candidate and then sets that, copying `src/Makevars.in` to `src/Makevars`.  Can you try that?

**Comment by ivan-pavlov** [comment]
Yes, I will try this.

One thing is unclear for me here.
If I want to run R script prior to building Rcpp code, how can I do that?

**Comment by eddelbuettel** [comment]
Put the code into a file named `configure`, make it executable.  It does not have to be made by `autoconf`.  Just use a first line `#!/usr/bin/env Rscript` and you are good to go. Try it!

**Comment by ivan-pavlov** [comment]
Ahh, alright
Thank you!

File: R/init.R
Comment (ivan-pavlov) [reply R/init.R]: Strangely, I can't run `.flatten()` (defined in `rhelpers.R`) function outside of `onAttach()`

**Comment by ivan-pavlov** [comment]
Sorry for taking so long, I was thinking I can finish vignette today, but I still need a few hours.

Strange thing's happening with `./configure` script.
I can successfully use it to build the package, but `R CMD check` fails with this message:
file `rvwgsoc.Rcheck/00install.out`
```
* installing *source* package ‘rvwgsoc’ ...
'Rscript' should not be used without a path -- see par. 1.6 of the manual
ERROR: configuration failed for package ‘rvwgsoc’
* removing ‘/Users/ivanpavlov/All_coding/Repos/GSoC2018/rvwgsoc.Rcheck/rvwgsoc’
```

I think this is because of `#!/usr/bin/env Rscript`. I was trying to fix this by running shell script to invoke R script:
```
#!/usr/bin/env bash

cmd="$(R RHOME)/bin/Rscript r_configure.R"
$cmd
```

But this doesn't help with `R CMD check`. 

Do you think it's possible to make `configure` script using `R` or should I try to rewrite it as a shell script?

File: R/init.R
Comment (eddelbuettel) [inline R/init.R]: Nice!

**Review by eddelbuettel — APPROVED** [review]
Looks good to me!

**Comment by eddelbuettel** [comment]
Not quite right in my situation as I have some headers missing, so I had to adjust `src/Makevars.in`.

Also:  

```
N  checking compilation flags in Makevars
   Package has both ‘src/Makevars.in’ and ‘src/Makevars’.
   Installation with --no-configure' is unlikely to work.  If you intended
   ‘src/Makevars’ to be used on Windows, rename it to ‘src/Makevars.win’
   otherwise remove it.  If ‘configure’ created ‘src/Makevars’, you need a
   ‘cleanup’ script.
```

I would add a script `cleanup` to  `rm -f src/Makevars src/*.o src/*.so src/*.dylib`.

**Comment by coatless** [comment]
Ivan, regarding the CRAN check error in `configure`, I would suggest using:

```bash
#! /bin/sh

# Note the shebang is using sh and not bash.

# Set R_HOME if not already set and switch into R's installation
: ${R_HOME:=`R RHOME`}

# Execute checks using R
cmd=`"${R_HOME}/bin/Rscript" r_configure.R`

$cmd
```

Remove any trace of `#!/usr/bin/env Rscript`

**Comment by coatless** [comment]
It looks like the unit tests are picking up an error:

```
Error: option '--data' cannot be specified more than once
```

I would double check the lists for 2 `--data` parameters.

c.f. 

```r
ext_train_data <- system.file("extdata", "binary_train.vw", package = "rvwgsoc")
test_vwmodel <- vwsetup()
vwtrain(test_vwmodel, data = ext_train_data)
# Starting VW training session
# VW v8.6.1
# Using data file: /Users/ronin/Google Drive/rprojects/tmp/rvwgsoc.Rcheck/rvwgsoc/extdata/binary_train.vw
# Using model file: /var/folders/b0/vt_1hj2d6yd8myx9lwh81pww0000gn/T//RtmpqqtYvW/vw_1534223196_mdl.vw
# Command line parameters: 
# -d /Users/ronin/Google Drive/rprojects/tmp/rvwgsoc.Rcheck/rvwgsoc/extdata/binary_train.vw -f /var/folders/b0/vt_1hj2d6yd8myx9lwh81pww0000gn/T//RtmpqqtYvW/vw_1534223196_mdl.vw --passes 1
# final_regressor = /var/folders/b0/vt_1hj2d6yd8myx9lwh81pww0000gn/T//RtmpqqtYvW/vw_1534223196_mdl.vw
# Num weight bits = 18
# learning rate = 0.5
# initial_t = 0
# power_t = 0.5
# Error: option '--data' cannot be specified more than once
```

**Comment by ivan-pavlov** [comment]
@coatless Thank you, I got rid of that CRAN check error.

Speaking about this unit test error with `Error: option '--data' cannot be specified more than once`

I think it's happening because of a whitespace character in the data path:
`-d /Users/ronin/Google Drive/rprojects/tmp/rvwgsoc.Rcheck/rvwgsoc/extdata/binary_train.vw`

This can be really confusing for Users, I've added a small check.

@eddelbuettel Thank you, I've added a `cleanup` script. 
For missing headers, should I also make adjustments in `src/Makevars.in`?

**Comment by eddelbuettel** [comment]
@coatless wrote

> Remove any trace of #!/usr/bin/env Rscript

Why?

**Comment by eddelbuettel** [comment]
@ivan-pavlov wrote

> For missing headers, should I also make adjustments in src/Makevars.in?

I would, and I would provide them. CRAN will use the Debian package on its Debian system.

**Comment by ivan-pavlov** [comment]
Yes, I will add them

**Comment by coatless** [comment]
@eddelbuettel  `#!/usr/bin/env Rscript` was triggering the error as it was included at the top of the R script. c.f.

https://github.com/wch/r-source/blob/5c2468dfde6c21aa0c9b362ebdc651315de31c1c/src/library/tools/R/check.R#L275-L282

**Comment by eddelbuettel** [comment]
Funny. I usually do `#!/usr/bin/env r` and I guess `littler` is not famous enough to warrant its own test.

In _this case_ the test is bullshit. We use R as a shell, awk, ... replacement just to test for files _and have no version dependency on R_.  Your sugggestion of using bash to call Rscript is by the book but still silly.  Could just do the whole script in bash then.

**Comment by ivan-pavlov** [comment]
I used R for `configure` script because I know R better than shell scripts and could prepare it faster and thought there’s no big difference what to use. There is a big difference after all, so I can rewrite this script in bash now.

**Comment by eddelbuettel** [comment]
... and you broke `configure`.  Do you actually test the code you upload?

```sh
edd@rob:~/git/rvwgsoc(feature/new-learning-modes)$ ./configure
./configure: 5: ./configure: PWD: not found
Fatal error: cannot open file '/tools/r_configure.R': No such file or directory
edd@rob:~/git/rvwgsoc(feature/new-learning-modes)$
```

Fixed in 73f5c3c03c07fd3ff302207e0b51d752776001fb

And don't rewrite it now.

**Comment by ivan-pavlov** [comment]
Thank you, I've accidentally used the wrong brackets.

Yes, I always test before uploading. Strange, but it's working even with the wrong brackets.

**Comment by eddelbuettel** [comment]
What's your shell?  Mine is bash, and it needs `{}` there (or alternative `$(pwd)` or `pwd` (using backticks I cannot how here); the latter two are equivalent and farm out a call to program `pwd` rather than the variable you call (which is better)).

**Comment by eddelbuettel** [comment]
Repo / code looks good, though I didn't have time for a detailed look.  

Let @coatless chime in and consider merging then.  

Also consider removing all the stale branches.  Whatever have been merged may as well be deleted.

Then ... write-up / vignette ?

**Comment by ivan-pavlov** [comment]
Thank you, I understand now. Mine is GNU bash, version 3.2.57, but yes, I was intending to use a variable and accidentally chose the wrong ones.

Vignette is my top priority right now, but there is a couple of small features I am thinking about.
But we should discuss them later probably.

**Comment by coatless** [comment]
@ivan-pavlov this looks good! Feel free to merge away.

In the next PR, could you added some defense if spaces are in the file path? 

Outside of that, please start focusing on documentation for the project and ensuring that CRAN checks are respected. In particular, I would recommend creating a vignette.

**Comment by ivan-pavlov** [comment]
Yes, I will add a defense for all file paths.

I am currently focused on documentation and especially a vignette, but I've also found a small bug in arguments parser and I am fixing it right now.

**Pull request merged**
